src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java

```diff
@@ -0,0 +1,415 @@
+            HttpHeader.SET_COOKIE2);
+    private static final EnumSet<HttpHeader> IGNORED_HEADERS = EnumSet.of(HttpHeader.CONNECTION, HttpHeader.KEEP_ALIVE,
+        HttpHeader.PROXY_CONNECTION, HttpHeader.TRANSFER_ENCODING, HttpHeader.UPGRADE);
+    private static final PreEncodedHttpField TE_TRAILERS = new PreEncodedHttpField(HttpHeader.TE, "trailers");
+    private static final PreEncodedHttpField C_SCHEME_HTTP = new PreEncodedHttpField(HttpHeader.C_SCHEME, "http");
+    private static final PreEncodedHttpField C_SCHEME_HTTPS = new PreEncodedHttpField(HttpHeader.C_SCHEME, "https");
+    private static final EnumMap<HttpMethod, PreEncodedHttpField> C_METHODS = new EnumMap<>(HttpMethod.class);
+
+    static {
+        for (HttpStatus.Code code : HttpStatus.Code.values()) {
+            STATUSES[code.getCode()] = new PreEncodedHttpField(HttpHeader.C_STATUS, Integer.toString(code.getCode()));
+        }
+        for (HttpMethod method : HttpMethod.values()) {
+            C_METHODS.put(method, new PreEncodedHttpField(HttpHeader.C_METHOD, method.asString()));
+        }
+    }
+
+    private final HpackContext _context;
+    private final boolean _debug;
+    private int _remoteMaxDynamicTableSize;
+    private int _localMaxDynamicTableSize;
+    private int _maxHeaderListSize;
+    private int _headerListSize;
+    private boolean _validateEncoding = true;
+
+    public HpackEncoder() {
+        this(4096, 4096, -1);
+    }
+
+    public HpackEncoder(int localMaxDynamicTableSize) {
+        this(localMaxDynamicTableSize, 4096, -1);
+    }
+
+    public HpackEncoder(int localMaxDynamicTableSize, int remoteMaxDynamicTableSize) {
+        this(localMaxDynamicTableSize, remoteMaxDynamicTableSize, -1);
+    }
+
+    public HpackEncoder(int localMaxDynamicTableSize, int remoteMaxDynamicTableSize, int maxHeaderListSize) {
+        _context = new HpackContext(remoteMaxDynamicTableSize);
+        _remoteMaxDynamicTableSize = remoteMaxDynamicTableSize;
+        _localMaxDynamicTableSize = localMaxDynamicTableSize;
+        _maxHeaderListSize = maxHeaderListSize;
+        _debug = LOG.isDebugEnabled();
+    }
+
+    public int getMaxHeaderListSize() {
+        return _maxHeaderListSize;
+    }
+
+    public void setMaxHeaderListSize(int maxHeaderListSize) {
+        _maxHeaderListSize = maxHeaderListSize;
+    }
+
+    public HpackContext getHpackContext() {
+        return _context;
+    }
+
+    public void setRemoteMaxDynamicTableSize(int remoteMaxDynamicTableSize) {
+        _remoteMaxDynamicTableSize = remoteMaxDynamicTableSize;
+    }
+
+    public void setLocalMaxDynamicTableSize(int localMaxDynamicTableSize) {
+        _localMaxDynamicTableSize = localMaxDynamicTableSize;
+    }
+
+    public boolean isValidateEncoding() {
+        return _validateEncoding;
+    }
+
+    public void setValidateEncoding(boolean validateEncoding) {
+        _validateEncoding = validateEncoding;
+    }
+
+    public void encode(ByteBuffer buffer, MetaData metadata) {
```

Comment:
Change to not throw `HpackException` as it changes the signature of the method.
The body of the method has been changed to throw a runtime exception `IllegalStateException`.
I verified that exception thrown from these blocks get reported as `INVALID_ARGUMENT` (which grpc status of `3`) errors.

```
ServerReflectionInfo rpc failed. Error code: 3, message: org.eclipse.jetty.http2.hpack.HpackException$SessionException: Could not hpack encode POST{u=http://127.0.0.4:10500/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo,HTTP/2.0,h=11,cl=-1}, debug in
fo: {"created":"@1588585825.367944000","description":"Error received from peer ipv4:127.0.0.1:9091","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"org.eclipse.jetty.http2.hpack.HpackException$SessionException: Could not hpack encode POST{u=h
ttp://127.0.0.4:10500/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo,HTTP/2.0,h=11,cl=-1}","grpc_status":3}    
```